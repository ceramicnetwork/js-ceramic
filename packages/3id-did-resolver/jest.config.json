{
  "moduleFileExtensions": ["js", "json", "ts"],
  "testEnvironment": "node",
  "resolver": "jest-resolver-enhanced",
  "extensionsToTreatAsEsm": [".ts"],
  "globals": {
    "ts-jest": {
      "useESM": true
    }
  },
  "moduleNameMapper": {
    "^(\\.{1,2}/.*)\\.js$": "$1"
  },
  "testRegex": "(/__tests__/.*(\\.|/)(test|spec))\\.tsx?$"
}
