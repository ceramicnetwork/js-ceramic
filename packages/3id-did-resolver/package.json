{
  "name": "@ceramicnetwork/3id-did-resolver",
  "version": "0.5.4",
  "description": "DID Resolver for the 3ID method",
  "keywords": [
    "Ceramic",
    "DID",
    "identity",
    "Data"
  ],
  "author": "Joel Torstensson <oed3933@gmail.com>",
  "homepage": "",
  "license": "(Apache-2.0 OR MIT)",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib"
  ],
  "scripts": {
    "test": "./node_modules/.bin/jest --coverage",
    "build": "../../node_modules/.bin/tsc -p tsconfig.json",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean",
    "lint": "./node_modules/.bin/eslint ./src --ext .js,.jsx,.ts,.tsx",
    "clean": "rm -rf ./lib"
  },
  "dependencies": {
    "@ceramicnetwork/common": "^0.14.1",
    "@ceramicnetwork/docid": "^0.3.0",
    "cids": "^1.0.2",
    "cross-fetch": "^3.0.6",
    "uint8arrays": "^1.1.0"
  },
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/preset-env": "^7.8.4",
    "@babel/preset-typescript": "^7.9.0",
    "@types/bs58": "^4.0.1",
    "@types/events": "^3.0.0",
    "@types/node": "^13.13.15",
    "@typescript-eslint/eslint-plugin": "^4.6.0",
    "@typescript-eslint/parser": "^4.6.0",
    "babel-jest": "^25.1.0",
    "did-resolver": "^2.1.1",
    "eslint": "^6.8.0",
    "eslint-plugin-jest": "^23.8.2",
    "jest": "^26.6.3",
    "tmp-promise": "^2.0.2"
  },
  "jest": {
    "transformIgnorePatterns": [
      "/node_modules(?!/did-jwt)/"
    ]
  },
  "gitHead": "c58eb9659f5fdd976aca2db6465619c20ba30ff9"
}
