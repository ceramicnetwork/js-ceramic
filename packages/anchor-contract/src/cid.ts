import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

const SHA256_CODE = 0x12
const DAG_CBOR_CODE = 0x71
const V1_PROOF_TYPE = 'f(bytes32)'

export function createCidFromHexValue(value: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(value.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}
