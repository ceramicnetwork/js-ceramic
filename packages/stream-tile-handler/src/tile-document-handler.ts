import jsonpatch from 'fast-json-patch'
import cloneDeep from 'lodash.clonedeep'
import { applyAnchorCommit } from '@ceramicnetwork/stream-handler-common'
import { TileDocument } from '@ceramicnetwork/stream-tile'
import {
  AnchorStatus,
  CommitData,
  CommitType,
  Context,
  SignatureStatus,
  SignatureUtils,
  StreamConstructor,
  StreamHandler,
  StreamState,
  StreamUtils,
} from '@ceramicnetwork/common'
import { StreamID } from '@ceramicnetwork/streamid'
import { SchemaValidation } from './schema-utils.js'

function stringArraysEqual(arr1: Array<string>, arr2: Array<string>) {
  if (arr1.length != arr2.length) {
    return false
  }
  for (let i = 0; i < arr1.length; i++) {
    if (arr1[i] !== arr2[i]) {
      return false
    }
  }
  return true
}

/**
 * TileDocument stream handler implementation
 */
export class TileDocumentHandler implements StreamHandler<TileDocument> {
  private readonly _schemaValidator: SchemaValidation

  constructor() {
    this._schemaValidator = new SchemaValidation()
  }

  get type(): number {
    return TileDocument.STREAM_TYPE_ID
  }

  get name(): string {
    return TileDocument.STREAM_TYPE_NAME
  }

  get stream_constructor(): StreamConstructor<TileDocument> {
    return TileDocument
  }

  /**
   * Applies commit (genesis|signed|anchor)
   * @param commitData - Commit (with JWS envelope or anchor proof, if available and extracted before application)
   * @param context - Ceramic context
   * @param state - Document state
   */
  async applyCommit(
    commitData: CommitData,
    context: Context,
    state?: StreamState
  ): Promise<StreamState> {
    if (state == null) {
      // apply genesis
      return this._applyGenesis(commitData, context)
    }

    if (StreamUtils.isAnchorCommitData(commitData)) {
      return this._applyAnchor(commitData, state)
    }

    return this._applySigned(commitData, state, context)
  }

  /**
   * Applies genesis commit
   * @param commitData - Genesis commit
   * @param context - Ceramic context
   * @private
   */
  async _applyGenesis(commitData: CommitData, context: Context): Promise<StreamState> {
    const did = context.did
    if (!did) throw new Error(`DID is not set`)
    const payload = commitData.commit
    const isSigned = StreamUtils.isSignedCommitData(commitData)
    if (isSigned) {
      const streamId = await StreamID.fromGenesis('tile', commitData.commit)
      const { controllers } = payload.header
      await SignatureUtils.verifyCommitSignature(commitData, did, controllers[0], null, streamId)
    } else if (payload.data) {
      throw Error('Genesis commit with contents should always be signed')
    }

    if (!(payload.header.controllers && payload.header.controllers.length === 1)) {
      throw new Error('Exactly one controller must be specified')
    }

    const state: StreamState = {
      type: TileDocument.STREAM_TYPE_ID,
      content: payload.data || {},
      metadata: payload.header,
      signature: isSigned ? SignatureStatus.SIGNED : SignatureStatus.GENESIS,
      anchorStatus: AnchorStatus.NOT_REQUESTED,
      log: [{ cid: commitData.cid, type: CommitType.GENESIS }],
    }

    if (state.metadata.schema) {
      await this._schemaValidator.validateSchema(context.api, state.content, state.metadata.schema)
    }

    return state
  }

  /**
   * Applies signed commit
   * @param commitData - Signed commit
   * @param state - Document state
   * @param context - Ceramic context
   * @private
   */
  async _applySigned(
    commitData: CommitData,
    state: StreamState,
    context: Context
  ): Promise<StreamState> {
    const did = context.did
    if (!did) throw new Error(`DID is not set`)
    // Retrieve the payload
    const payload = commitData.commit
    StreamUtils.assertCommitLinksToState(state, payload)

    // Verify the signature
    const controller = state.next?.metadata?.controllers?.[0] || state.metadata.controllers[0]
    if (!controller) throw new Error(`Controller is not set`)
    const streamId = StreamUtils.streamIdFromState(state)
    await SignatureUtils.verifyCommitSignature(commitData, did, controller, null, streamId)

    if (payload.header.controllers) {
      if (payload.header.controllers.length !== 1) {
        throw new Error('Exactly one controller must be specified')
      }
      if (!payload.header.controllers[0]) {
        throw new Error('Controller cannot be updated to an undefined value.')
      }
    }

    if (
      state.metadata.forbidControllerChange &&
      payload.header.controllers &&
      !stringArraysEqual(payload.header.controllers, state.metadata.controllers)
    ) {
      const genesisLogEntry = state.log[0]
      const streamId = new StreamID(TileDocument.STREAM_TYPE_ID, genesisLogEntry.cid)
      throw new Error(
        `Cannot change controllers since 'forbidControllerChange' is set. Tried to change controllers for Stream ${streamId} from ${JSON.stringify(
          state.metadata.controllers
        )} to ${payload.header.controllers}`
      )
    }

    if (
      payload.header.forbidControllerChange !== undefined &&
      payload.header.forbidControllerChange !== state.metadata.forbidControllerChange
    ) {
      throw new Error("Changing 'forbidControllerChange' metadata property is not allowed")
    }

    const nextState = cloneDeep(state)

    nextState.signature = SignatureStatus.SIGNED
    nextState.anchorStatus = AnchorStatus.NOT_REQUESTED

    nextState.log.push({ cid: commitData.cid, type: CommitType.SIGNED })

    const oldContent = state.next?.content ?? state.content
    const oldMetadata = state.next?.metadata ?? state.metadata

    const newContent = jsonpatch.applyPatch(oldContent, payload.data).newDocument
    const newMetadata = { ...oldMetadata, ...payload.header }

    if (newMetadata.schema) {
      // TODO: SchemaValidation.validateSchema does i/o to load a Stream.  We should pre-load
      // the schema into the CommitData so that commit application can be a simple state
      // transformation with no i/o.
      await this._schemaValidator.validateSchema(context.api, newContent, newMetadata.schema)
    }

    nextState.next = {
      content: newContent,
      metadata: newMetadata,
    }

    return nextState
  }

  /**
   * Applies anchor commit
   * @param commitData - Anchor commit
   * @param state - Document state
   * @private
   */
  async _applyAnchor(commitData: CommitData, state: StreamState): Promise<StreamState> {
    return applyAnchorCommit(commitData, state)
  }
}
