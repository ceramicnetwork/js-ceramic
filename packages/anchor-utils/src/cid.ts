import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

export const SHA256_CODE = 0x12
export const KECCAK_256_CODE = 0x1b
export const DAG_CBOR_CODE = 0x71
export const ETH_TX_CODE = 0x93

export function createCidFromHexValue(hexValue: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor(address,bytes32)'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

const V1_PROOF_TYPE = 'f(bytes32)'

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}

/**
 * Converts ETH address to CID
 * @param hash - ETH hash
 */
export function convertEthHashToCid(hexValue: string): CID {
  const multihash = createMultihash(KECCAK_256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, ETH_TX_CODE, multihash)
}
