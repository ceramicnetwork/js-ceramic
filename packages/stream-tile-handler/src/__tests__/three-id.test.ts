import { jest } from '@jest/globals'
import { CID } from 'multiformats/cid'
import { decode as decodeMultiHash } from 'multiformats/hashes/digest'
import * as dagCBOR from '@ipld/dag-cbor'
import * as KeyDidResolver from 'key-did-resolver'
import { wrapDocument } from '@ceramicnetwork/3id-did-resolver'
import type { DID } from 'dids'
import {
  CeramicApi,
  CommitData,
  CommitType,
  Context,
  IpfsApi,
  SignedCommitContainer,
  TestUtils,
} from '@ceramicnetwork/common'
import { TileDocumentHandler } from '../tile-document-handler.js'
import { TileDocument } from '@ceramicnetwork/stream-tile'
import cloneDeep from 'lodash.clonedeep'
import * as sha256 from '@stablelib/sha256'
import * as uint8arrays from 'uint8arrays'

jest.unstable_mockModule('did-jwt', () => {
  return {
    // TODO - We should test for when this function throws as well
    // Mock: Blindly accept a signature
    verifyJWS: (): void => {
      return
    },
    // And these functions are required for the test to run ¯\_(ツ)_/¯
    resolveX25519Encrypters: () => {
      return []
    },
    createJWE: () => {
      return {}
    },
  }
})

const hash = (data: string): CID =>
  CID.create(
    1,
    0x12,
    decodeMultiHash(
      uint8arrays.fromString('1220' + sha256.hash(uint8arrays.fromString(data)), 'base16')
    )
  )

const FAKE_CID_1 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')

const COMMITS = {
  genesis: {
    header: {
      tags: ['3id'],
      controllers: ['did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV'],
    },
    data: { publicKeys: { test: '0xabc' } },
  },
  genesisGenerated: {
    jws: {
      payload: 'bbbb',
      signatures: [
        {
          protected:
            'eyJraWQiOiJkaWQ6a2V5OnpRM3Nod3NDZ0ZhbkJheDZVaWFMdTFvR3ZNN3ZodXFvVzg4VkJVaVVUQ2VIYlRlVFYiLCJhbGciOiJFUzI1NksifQ',
          signature: 'cccc',
        },
      ],
      link: CID.parse('bafyreig7dosektvux6a55mphri6piy3g5k4otxinanfevjeehix6rqfeym'),
    },
    linkedBlock: {
      data: {
        publicKeys: {
          test: '0xabc',
        },
      },
      header: {
        controllers: ['did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV'],
        tags: ['3id'],
      },
    },
  },
  r1: {
    desiredContent: { publicKeys: { test: '0xabc' }, other: 'data' },
    commit: {
      jws: {
        payload: 'bbbb',
        signatures: [
          {
            protected:
              'eyJraWQiOiJkaWQ6a2V5OnpRM3Nod3NDZ0ZhbkJheDZVaWFMdTFvR3ZNN3ZodXFvVzg4VkJVaVVUQ2VIYlRlVFYiLCJhbGciOiJFUzI1NksifQ',
            signature: 'cccc',
          },
        ],
        link: 'bafyreib2rxk3rybk3aobmv5cjuql3bm2twh4jo5uxgf5kpqcsgz7soitae',
      },
      payload: {
        id: 'bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu',
        data: [
          {
            op: 'add',
            path: '/other',
            value: 'data',
          },
        ],
        prev: 'bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu',
        header: {},
      },
    },
  },
  r2: { commit: { proof: FAKE_CID_4, id: FAKE_CID_1, prev: FAKE_CID_2 } },
  proof: {
    blockNumber: 123456,
    blockTimestamp: 1615799679,
    chainId: 'fakechain:123',
  },
}

const serialize = (data: any): any => {
  if (Array.isArray(data)) {
    const serialized = []
    for (const item of data) {
      serialized.push(serialize(item))
    }
    return serialized
  }
  const cid = CID.asCID(data)
  if (!cid && typeof data === 'object') {
    const serialized: Record<string, any> = {}
    for (const prop in data) {
      serialized[prop] = serialize(data[prop])
    }
    return serialized
  }
  if (cid) {
    return data.toString()
  }
  return data
}

let did: DID
let tileDocumentHandler: TileDocumentHandler
let context: Context

beforeAll(async () => {
  const recs: Record<string, any> = {}
  const ipfs = {
    dag: {
      put(rec: any, cid?: CID): any {
        if (cid) {
          recs[cid.toString()] = { value: rec }
          return cid
        }
        // stringify as a way of doing deep copy
        const clone = cloneDeep(rec)
        const c = hash(JSON.stringify(clone))
        recs[c.toString()] = { value: clone }
        return c
      },
      get(cid: any): any {
        return recs[cid.toString()]
      },
    },
  } as IpfsApi

  const threeIdResolver = {
    '3': async (did) => ({
      didResolutionMetadata: { contentType: 'application/did+json' },
      didDocument: wrapDocument(
        {
          publicKeys: {
            signing: 'zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV',
            encryption: 'z6LSfQabSbJzX8WAm1qdQcHCHTzVv8a2u6F7kmzdodfvUCo9',
          },
        },
        did
      ),
      didDocumentMetadata: {},
    }),
  }

  const { DID } = await import('dids')
  const keyDidResolver = KeyDidResolver.getResolver()
  did = new DID({
    resolver: {
      ...threeIdResolver,
      ...keyDidResolver,
    },
  })
  did.createJWS = jest.fn(async () => {
    // fake jws
    return {
      payload: 'bbbb',
      signatures: [
        {
          protected:
            'eyJraWQiOiJkaWQ6a2V5OnpRM3Nod3NDZ0ZhbkJheDZVaWFMdTFvR3ZNN3ZodXFvVzg4VkJVaVVUQ2VIYlRlVFYiLCJhbGciOiJFUzI1NksifQ',
          signature: 'cccc',
        },
      ],
    }
  })
  ;(did as any)._id = 'did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV'
  const api = {
    getSupportedChains: jest.fn(async () => {
      return ['fakechain:123']
    }),
    did,
  }

  context = {
    did,
    ipfs,
    anchorService: null,
    api: api as unknown as CeramicApi,
  }
})

beforeEach(async () => {
  tileDocumentHandler = new TileDocumentHandler()
})

it('is constructed correctly', async () => {
  expect(tileDocumentHandler.name).toEqual('tile')
})

it('makes genesis commit correctly', async () => {
  const commit = (await TileDocument.makeGenesis(context.api, COMMITS.genesis.data, {
    ...COMMITS.genesis.header,
  })) as SignedCommitContainer
  const { jws, linkedBlock } = commit

  const payload = dagCBOR.decode(linkedBlock)
  const generated = { jws: serialize(jws), linkedBlock: serialize(payload) }

  // Add the 'unique' header field to the data used to generate the expected genesis commit
  const genesis = cloneDeep(COMMITS.genesis)
  genesis.header['unique'] = generated.linkedBlock.header.unique

  const expected = await did.createDagJWS(genesis)
  expect(expected).toBeDefined()

  const { jws: eJws, linkedBlock: eLinkedBlock } = expected
  const ePayload = dagCBOR.decode(eLinkedBlock)
  const signed = { jws: serialize(eJws), linkedBlock: serialize(ePayload) }

  expect(generated).toEqual(serialize(signed))
})

it('applies genesis commit correctly', async () => {
  const commit = (await TileDocument.makeGenesis(context.api, COMMITS.genesis.data, {
    controllers: [did.id],
    tags: ['3id'],
  })) as SignedCommitContainer
  await context.ipfs.dag.put(commit, FAKE_CID_1)

  const payload = dagCBOR.decode(commit.linkedBlock)
  await context.ipfs.dag.put(payload, commit.jws.link)

  const signedCommitData = {
    cid: FAKE_CID_1,
    type: CommitType.GENESIS,
    commit: payload,
    envelope: commit.jws,
  }
  const streamState = await tileDocumentHandler.applyCommit(signedCommitData, context)
  delete streamState.metadata.unique
  expect(streamState).toMatchSnapshot()
})

it('makes signed commit correctly', async () => {
  await context.ipfs.dag.put(COMMITS.genesisGenerated.jws, FAKE_CID_1)
  await context.ipfs.dag.put(
    COMMITS.genesisGenerated.linkedBlock,
    COMMITS.genesisGenerated.jws.link
  )

  const genesisCommitData = {
    cid: FAKE_CID_1,
    type: CommitType.GENESIS,
    commit: COMMITS.genesisGenerated.linkedBlock,
    envelope: COMMITS.genesisGenerated.jws,
  }
  const state = await tileDocumentHandler.applyCommit(genesisCommitData, context)
  const state$ = TestUtils.runningState(state)
  const doc = new TileDocument(state$, context)

  await expect(doc.makeCommit({} as CeramicApi, COMMITS.r1.desiredContent)).rejects.toThrow(
    /No DID/
  )

  const commit = (await doc.makeCommit(
    context.api,
    COMMITS.r1.desiredContent
  )) as SignedCommitContainer
  const { jws: rJws, linkedBlock: rLinkedBlock } = commit
  const rPayload = dagCBOR.decode(rLinkedBlock)
  expect({ jws: serialize(rJws), payload: serialize(rPayload) }).toEqual(COMMITS.r1.commit)
})

it('applies signed commit correctly', async () => {
  const genesisCommit = (await TileDocument.makeGenesis(context.api, COMMITS.genesis.data, {
    controllers: [did.id],
    tags: ['3id'],
  })) as SignedCommitContainer
  await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

  const payload = dagCBOR.decode(genesisCommit.linkedBlock)
  await context.ipfs.dag.put(payload, genesisCommit.jws.link)

  // apply genesis
  const genesisCommitData = {
    cid: FAKE_CID_1,
    type: CommitType.GENESIS,
    commit: payload,
    envelope: genesisCommit.jws,
  }
  let state = await tileDocumentHandler.applyCommit(genesisCommitData, context)

  const state$ = TestUtils.runningState(state)
  const doc = new TileDocument(state$, context)
  const signedCommit = (await doc.makeCommit(
    context.api,
    COMMITS.r1.desiredContent
  )) as SignedCommitContainer

  await context.ipfs.dag.put(signedCommit, FAKE_CID_2)

  const sPayload = dagCBOR.decode(signedCommit.linkedBlock)
  await context.ipfs.dag.put(sPayload, signedCommit.jws.link)

  // apply signed
  const signedCommitData = {
    cid: FAKE_CID_2,
    type: CommitType.SIGNED,
    commit: sPayload,
    envelope: signedCommit.jws,
  }
  state = await tileDocumentHandler.applyCommit(signedCommitData, context, state)
  delete state.metadata.unique
  delete state.next.metadata.unique
  expect(state).toMatchSnapshot()
})

it('throws error if commit signed by wrong DID', async () => {
  const genesisCommit = (await TileDocument.makeGenesis(context.api, COMMITS.genesis.data, {
    controllers: ['did:3:fake'],
    tags: ['3id'],
  })) as SignedCommitContainer
  await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

  const payload = dagCBOR.decode(genesisCommit.linkedBlock)
  await context.ipfs.dag.put(payload, genesisCommit.jws.link)

  const genesisCommitData = {
    cid: FAKE_CID_1,
    type: CommitType.GENESIS,
    commit: payload,
    envelope: genesisCommit.jws,
  }
  await expect(tileDocumentHandler.applyCommit(genesisCommitData, context)).rejects.toThrow(
    /invalid_jws: not a valid verificationMethod for issuer/
  )
})

it('applies anchor commit correctly', async () => {
  const genesisCommit = (await TileDocument.makeGenesis(context.api, COMMITS.genesis.data, {
    controllers: [did.id],
    tags: ['3id'],
  })) as SignedCommitContainer
  await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

  const payload = dagCBOR.decode(genesisCommit.linkedBlock)
  await context.ipfs.dag.put(payload, genesisCommit.jws.link)

  // apply genesis
  const genesisCommitData = {
    cid: FAKE_CID_1,
    type: CommitType.GENESIS,
    commit: payload,
    envelope: genesisCommit.jws,
  }
  let state = await tileDocumentHandler.applyCommit(genesisCommitData, context)

  const state$ = TestUtils.runningState(state)
  const doc = new TileDocument(state$, context)
  const signedCommit = (await doc.makeCommit(
    context.api,
    COMMITS.r1.desiredContent
  )) as SignedCommitContainer

  await context.ipfs.dag.put(signedCommit, FAKE_CID_2)

  const sPayload = dagCBOR.decode(signedCommit.linkedBlock)
  await context.ipfs.dag.put(sPayload, signedCommit.jws.link)

  // apply signed
  const signedCommitData = {
    cid: FAKE_CID_2,
    type: CommitType.SIGNED,
    commit: sPayload,
    envelope: signedCommit.jws,
  }
  state = await tileDocumentHandler.applyCommit(signedCommitData, context, state)

  await context.ipfs.dag.put(COMMITS.proof, FAKE_CID_4)
  // apply anchor
  const anchorCommitData = {
    cid: FAKE_CID_3,
    type: CommitType.ANCHOR,
    commit: COMMITS.r2.commit,
    proof: COMMITS.proof,
  } as CommitData
  state = await tileDocumentHandler.applyCommit(anchorCommitData, context, state)
  delete state.metadata.unique
  expect(state).toMatchSnapshot()
})
