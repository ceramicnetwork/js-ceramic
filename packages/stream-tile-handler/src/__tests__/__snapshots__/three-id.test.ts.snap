// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applies anchor commit correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
    "blockTimestamp": 1615799679,
    "chainId": "fakechain:123",
  },
  "anchorStatus": 3,
  "content": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "timestamp": 1615799679,
      "type": 0,
    },
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "timestamp": 1615799679,
      "type": 1,
    },
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "timestamp": 1615799679,
      "type": 2,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    ],
    "tags": Array [
      "3id",
    ],
  },
  "signature": 2,
  "type": 0,
}
`;

exports[`applies genesis commit correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    ],
    "tags": Array [
      "3id",
    ],
  },
  "signature": 2,
  "type": 0,
}
`;

exports[`applies signed commit correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    ],
    "tags": Array [
      "3id",
    ],
  },
  "next": Object {
    "content": Object {
      "other": "data",
      "publicKeys": Object {
        "test": "0xabc",
      },
    },
    "metadata": Object {
      "controllers": Array [
        "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
      ],
      "tags": Array [
        "3id",
      ],
    },
  },
  "signature": 2,
  "type": 0,
}
`;
